app/controllers/CompilerService.scala

```diff
@@ -58,11 +58,33 @@ object CompilerService extends Controller {
 
   protected[controllers] type ModelMaker = (String, List[Widget], String) => ModelResultV
   protected[controllers] val modelFromURL: ModelMaker   = (url,   _, hostUri) =>
-    fetchURL(url, hostUri) map (CompiledModel.fromNlogoContents(_))
+    fetchURL(url, hostUri) flatMap compileNlogo
   protected[controllers] val modelFromCode:  ModelMaker = (code,  widgets, _) =>
     CompiledModel.fromModel(Model(code, widgets)).successNel
   protected[controllers] val modelFromNlogo: ModelMaker = (nlogo, _, _)       =>
-    CompiledModel.fromNlogoContents(nlogo).successNel
+    compileNlogo(nlogo)
+
+  private def compileNlogo(nlogo: String): ModelResultV =
+    stringifyNonCompilerExceptions(CompiledModel.fromNlogoContents(nlogo))
+
+  private def stringifyNonCompilerExceptions(v: ValidationNel[Exception, CompiledModel]): ModelResultV =
+    v.fold(
+      { failures =>
+        val firstFailure = failures.head match {
+          case e: CompilerException => \/.right(NonEmptyList(e))
+          case e: Exception         => \/.left(NonEmptyList(e))
+        }
+        failures.tail.foldLeft[\/[NonEmptyList[Exception], NonEmptyList[CompilerException]]](firstFailure) {
+          case (ces: \/-[NonEmptyList[CompilerException]], ce: CompilerException) => ces.map(_.<::(ce))
+          case (ces: \/-[NonEmptyList[CompilerException]], e: Exception)          => \/.left(NonEmptyList(e))
+          case (es: -\/[NonEmptyList[Exception]],          e: CompilerException)  => es
+          case (es: -\/[NonEmptyList[Exception]],          e: Exception)          => es.leftMap(_.<::(e))
+        }.fold(
+            otherExceptions    => otherExceptions.map(_.getMessage).failure,
+            compilerExceptions => compilerExceptions.failure.successNel[String])
+      },
```

Comment:
I believe the operation in the preceding 12 lines can be done in fewer lines using casts. Perhaps there's an easier way in scalaz? Basically, we're looking to return any CompilerExceptions while converting any non-compiler-exceptions into strings.